base/regex.jl

```diff
@@ -363,6 +363,9 @@ macro s_str(string) SubstitutionString(string) end
 replace_err(repl) = error("Bad replacement string: $repl")
 
 function _write_capture(io, re, group)
+    # in the case of an optional group that doesn't match, don't write anything
+    # e.g. replace("The fox.", r"fox(es)?" => s"bus\1")
+    isa(re.match_data, Ptr{Nothing}) && return
```

Comment:
This doesn't look right --- isn't `match_data` always a `Ptr`?

We might have to modify `substring_length_bynumber` to return 0 in this error case, but I'm not sure whether that's the best way to do it.

Reply:
Ok yes. Looking at the PCRE2 doc for substring (https://code.woboq.org/qt5/qtbase/src/3rdparty/pcre2/src/pcre2_substring.c.html) L315 onwards, it returns a PCRE2_ERROR_UNSET code (-55) when it can't find a match. 

So locally I did the following:

1. modify `substring_length_bynumber` as suggested with

```julia
function substring_length_bynumber(match_data, number)
    s = Base.RefValue{Csize_t}()
    rc = ccall((:pcre2_substring_length_bynumber_8, PCRE_LIB), Cint,
          (Ptr{Cvoid}, UInt32, Ref{Csize_t}), match_data, number, s)
    if rc < 0
        rc == -55 && return 0
        error("PCRE error: $(err_message(rc))")
    end
    convert(Int, s[])
end
```

2. modify `_write_capture` so that it doesn't go further when it has a length of 0 (otherwise the next PCRE function fails):

```julia
function _write_capture2(io, re, group)
    len = substring_length_bynumber(re.match_data, group)
    len == 0 && return
    ensureroom(io, len+1)
    PCRE.substring_copy_bynumber(re.match_data, group,
        pointer(io.data, io.ptr), len+1)
    io.ptr += len
    io.size = max(io.size, io.ptr - 1)
end
```

This seems to work including with multiple captures:

```julia-repl
julia> replace("The quick fox quickly.", r"(quick)?\sfox(es)?\s(run)?" => s"\1 bus\2 \3")
"The quick bus quickly."
```

Should I try that in the PR or should I close it and let someone who knows more about PCRE take over?